src/posit/connect/users.py

```diff
@@ -13,18 +11,90 @@
-    guid: str
-    email: str
-    username: str
-    first_name: str
-    last_name: str
-    user_role: str
-    created_time: datetime
-    updated_time: datetime
-    active_time: datetime
-    confirmed: bool
-    locked: bool
+class User(dict):
+
+    @property
+    def guid(self) -> str:
+        return self.get("guid")  # type: ignore
+
+    @property
+    def email(self) -> str:
+        return self.get("email")  # type: ignore
+
+    @property
+    def username(self) -> str:
+        return self.get("username")  # type: ignore
+
+    @property
+    def first_name(self) -> str:
+        return self.get("first_name")  # type: ignore
+
+    @property
+    def last_name(self) -> str:
+        return self.get("last_name")  # type: ignore
+
+    @property
+    def user_role(self) -> str:
+        return self.get("user_role")  # type: ignore
+
+    @property
+    def created_time(self) -> str:
+        return self.get("created_time")  # type: ignore
+
+    @property
+    def updated_time(self) -> str:
+        return self.get("updated_time")  # type: ignore
+
+    @property
+    def active_time(self) -> str:
+        return self.get("active_time")  # type: ignore
+
+    @property
+    def confirmed(self) -> bool:
+        return self.get("confirmed")  # type: ignore
+
+    @property
+    def locked(self) -> bool:
+        return self.get("locked")  # type: ignore
+
+    def __setattr__(self, name: str, value: Any) -> None:
+        raise AttributeError("Cannot set attributes: use update() instead.")
+
+    def _update(self, body):
+        self.get("session").patch(self.get("url"), json=body)
+        # If the request is successful, update the local object
+        super().update(body)
+        # TODO(#99): that patch request returns a payload on success,
+        # so we should instead update the local object with that payload
+        # (includes updated_time)
+
+    def update(  # type: ignore
+        self,
+        # Not all properties are settable, so we enumerate them here
+        # (also for type-hinting purposes)
+        email: Optional[str] = None,
+        username: Optional[str] = None,
+        first_name: Optional[str] = None,
+        last_name: Optional[str] = None,
+        user_role: Optional[str] = None,
+        # TODO(#100): in the API, this goes via POST /v1/users/{guid}/lock
+        # accept it here and make that request? Or add a .lock() method?
+        # locked: Optional[bool] = None,
+    ) -> None:
+        kwargs = {}
+        if email is not None:
+            kwargs["email"] = email
+        if username is not None:
+            kwargs["username"] = username
+        if first_name is not None:
+            kwargs["first_name"] = first_name
+        if last_name is not None:
+            kwargs["last_name"] = last_name
+        if user_role is not None:
+            kwargs["user_role"] = user_role
+        # if locked is not None:
+        #     kwargs["locked"] = locked
+        self._update(kwargs)
```

Comment:
This looks good. But should we enforce these logical constraints or defer this behavior to the server? In other words, how do we handle behavioral changes to the PATCH endpoint between versions of Connect?